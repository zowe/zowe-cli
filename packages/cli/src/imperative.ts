/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as path from "path";

import { IImperativeConfig } from "@zowe/imperative";
import { Constants } from "./Constants";

import { ZosmfProfile } from "@zowe/zosmf-for-zowe-sdk";
import { ZosTsoProfile } from "@zowe/zos-tso-for-zowe-sdk";
import { ZosUssProfile } from "@zowe/zos-uss-for-zowe-sdk";
import { ProfileConstants } from "@zowe/core-for-zowe-sdk";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: [
        "daemon/*.definition!(.d).*s",
        "provisioning/*.definition!(.d).*s",
        "workflows/*.definition!(.d).*s",
        "zosconsole/*.definition!(.d).*s",
        "zosfiles/*.definition!(.d).*s",
        "zosjobs/*.definition!(.d).*s",
        "zosmf/*.definition!(.d).*s",
        "zostso/*.definition!(.d).*s",
        "zosuss/*.definition!(.d).*s",
        "zoslogs/*.definition!(.d).*s"
    ],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    webHelpLogoImgPath: path.join(__dirname, "..", "web-help-logo.png"),
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        }
    },
    apimlConnLookup: [
        {
            apiId: "ibm.zosmf",
            gatewayUrl: "api/v1",
            connProfType: "zosmf"
        }
    ],
    baseProfile: {
        ...ProfileConstants.BaseProfile,
        authConfig: [
            {
                serviceName: "apiml",
                handler: __dirname + "/auth/ApimlAuthHandler",
                login: {
                    summary: ProfileConstants.APIML_LOGIN_SUMMARY,
                    description: ProfileConstants.APIML_LOGIN_DESCRIPTION,
                    examples: [
                        ProfileConstants.APIML_LOGIN_EXAMPLE1,
                        ProfileConstants.APIML_LOGIN_EXAMPLE2
                    ],
                    options: [
                        ProfileConstants.BASE_OPTION_HOST,
                        ProfileConstants.BASE_OPTION_PORT,
                        ProfileConstants.BASE_OPTION_USER,
                        ProfileConstants.BASE_OPTION_PASSWORD,
                        ProfileConstants.BASE_OPTION_REJECT_UNAUTHORIZED,
                        ProfileConstants.BASE_OPTION_CERT_FILE,
                        ProfileConstants.BASE_OPTION_CERT_KEY_FILE,
                        ProfileConstants.BASE_OPTION_CERT_ACCOUNT,
                        ProfileConstants.BASE_OPTION_CERT_KEY_ACCOUNT
                    ]
                },
                logout: {
                    summary: ProfileConstants.APIML_LOGOUT_SUMMARY,
                    description: ProfileConstants.APIML_LOGOUT_DESCRIPTION,
                    examples: [
                        ProfileConstants.APIML_LOGOUT_EXAMPLE1,
                        ProfileConstants.APIML_LOGOUT_EXAMPLE2
                    ],
                    options: [
                        ProfileConstants.BASE_OPTION_HOST,
                        ProfileConstants.BASE_OPTION_PORT,
                        ProfileConstants.APIML_LOGOUT_OPTION_TOKEN_TYPE,
                        ProfileConstants.BASE_OPTION_TOKEN_VALUE,
                        ProfileConstants.BASE_OPTION_REJECT_UNAUTHORIZED
                    ]
                }
            }
        ]
    },
    authGroupConfig: {
        authGroup: {
            summary: ProfileConstants.AUTH_GROUP_SUMMARY,
            description: ProfileConstants.AUTH_GROUP_DESCRIPTION
        }
    },
    configAutoInitCommandConfig: {
        handler: __dirname + "/config/auto-init/ApimlAutoInitHandler",
        provider: "APIML",
        autoInit: {
            options: [
                ProfileConstants.AUTO_INIT_OPTION_HOST,
                ProfileConstants.AUTO_INIT_OPTION_PORT,
                ProfileConstants.AUTO_INIT_OPTION_USER,
                ProfileConstants.AUTO_INIT_OPTION_PASSWORD,
                ProfileConstants.AUTO_INIT_OPTION_REJECT_UNAUTHORIZED,
                ProfileConstants.AUTO_INIT_OPTION_TOKEN_TYPE,
                ProfileConstants.AUTO_INIT_OPTION_TOKEN_VALUE,
                ProfileConstants.AUTO_INIT_OPTION_CERT_FILE,
                ProfileConstants.AUTO_INIT_OPTION_CERT_KEY_FILE,
                ProfileConstants.BASE_OPTION_CERT_ACCOUNT,
                ProfileConstants.BASE_OPTION_CERT_KEY_ACCOUNT
            ]
        },
        profileType: "base"
    },
    profiles: [
        ZosmfProfile,
        ZosTsoProfile,
        ZosUssProfile
    ]
};
module.exports = config;
