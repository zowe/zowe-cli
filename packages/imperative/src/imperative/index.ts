/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./env/EnvironmentalVariableSettings";
export * from "./doc/IApimlSvcAttrs";
export * from "./doc/IDaemonContext";
export * from "./doc/IImperativeEnvironmentalVariableSetting";
export * from "./doc/IImperativeEnvironmentalVariableSettings";
export * from "./doc/IImperativeConfig";
export * from "./doc/IImperativeLoggingConfig";
export * from "./ConfigurationLoader";
export * from "./ConfigurationValidator";
export * from "./DefinitionTreeResolver";
export * from "./Imperative";
export * from "./LoggingConfigurer";
export * from "./plugins/AbstractPluginLifeCycle";
export * from "./plugins/PluginManagementFacility";
export * from "./auth/doc/IAuthHandlerApi";
export * from "./auth/handlers/AbstractAuthHandler";
export * from "./auth/handlers/BaseAuthHandler";
export * from "./config/cmd/auto-init/handlers/BaseAutoInitHandler";
