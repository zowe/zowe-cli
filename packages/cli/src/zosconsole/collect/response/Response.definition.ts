/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandDefinition } from "@zowe/imperative";
import { Constants } from "../../../Constants";

export const SyncResponseCommandDefinition: ICommandDefinition = {
    name: "sync-responses",
    aliases: ["sr"],
    summary: "Collect outstanding synchronous console response messages",
    description: `The z/OSMF console REST APIs return a "solicited response key" after successfully issuing ` +
    `a synchronous console command that produces solicited responses. You can use the "solicited response key"` +
    `on the "sync-responses" command to collect any additional outstanding solicited responses ` +
    `from the console the command was issued.` +
    `\n\n` +
    `In general, when issuing a z/OS console command, z/OS applications route responses to the originating console. ` +
    `The command response messages are referred to as "solicited command responses" (i.e. direct responses to the command issued). ` +
    `When issuing a z/OS console command using ${Constants.DISPLAY_NAME}, collection of all solicited command responses ` +
    `is attempted by default. However, there is no z/OS mechanism that indicates the total number of response messages that may be produced ` +
    `from a given command. Therefore, the ${Constants.DISPLAY_NAME} console APIs return a "solicited response key" ` +
    `that can be used to "follow-up" and collect any additional solicited command responses.`,
    type: "command",
    handler: __dirname + "/Response.handler",
    profile: {
        optional: ["zosmf"]
    },
    positionals: [
        {
            name: "responsekey",
            type: "string",
            description: `The "solicited response key" provided in response to a previously issued console command. ` +
            `Used by the z/OSMF console API to collect any additional outstanding solicited responses from a ` +
            `previously issued console command.`,
            regex: "^[a-zA-Z0-9]+$",
            required: true
        }
    ],
    options: [
        {
            name: "console-name",
            aliases: ["cn", "c"],
            description: `The name of the z/OS extended MCS console to direct the command. ` +
            `You must have the required authority to access the console specified. ` +
            `You may also specify an arbitrary name, if your installation allows dynamic creation ` +
            `of consoles with arbitrary names.`,
            type: "string",
            allowableValues: {
                values: [/^[a-zA-Z0-9]+$/.source],
                caseSensitive: false
            }
        }
    ],
    examples: [
        {
            description: "Collect any outstanding additional solicited response messages",
            options: "C4866969"
        }
    ]
};
