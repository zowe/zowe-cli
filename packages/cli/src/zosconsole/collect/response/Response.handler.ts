/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IHandlerParameters } from "@zowe/imperative";
import { CollectCommand, ICollectParms, IConsoleResponse } from "@zowe/zos-console-for-zowe-sdk";
import { ZosmfBaseHandler } from "@zowe/zosmf-for-zowe-sdk";

/**
 * Handle to collect a MVS console command response
 * @export
 * @class Handler
 * @implements {ICommandHandler}
 */
export default class Handler extends ZosmfBaseHandler {

    public async processCmd(commandParameters: IHandlerParameters) {

        const collectParms: ICollectParms = {
            commandResponseKey: commandParameters.arguments.responsekey,
            consoleName: commandParameters.arguments["console-name"]
        };

        const response: IConsoleResponse = await CollectCommand.collect(this.mSession, collectParms);

        // Print out the response
        commandParameters.response.console.log(response.commandResponse);

        // Return as an object when using --response-format-json
        commandParameters.response.data.setObj(response);

    }
}
