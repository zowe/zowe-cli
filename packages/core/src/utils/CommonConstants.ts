/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IMessageDefinition, apiErrorHeader } from "npm:@zowe/imperative";

/**
 * Error message that no z/OSMF version parameter string was supplied.
 * @static
 * @type {IMessageDefinition}
 * @memberof CommonConstants
 */
export const nozOSMFVersion: IMessageDefinition = {
    message: apiErrorHeader.message + ` No z/OSMF version parameter was supplied.`
};


/**
 * Error message that no session was supplied.
 * @static
 * @type {IMessageDefinition}
 * @memberof CommonConstants
 */
export const noSession: IMessageDefinition = {
    message: apiErrorHeader.message + ` No session was supplied.`
};
