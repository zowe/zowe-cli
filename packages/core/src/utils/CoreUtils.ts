/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as fs from "fs";
import { IImperativeError, Logger, TextUtils } from "@zowe/imperative";

const stringWidth = require("string-width");

/**
 * Adds padding in case you want a longer string.
 *
 * @param {string} str              - string to add the padding to
 * @param {number} len              - final length of the padded string
 * @param {string} [padChar=" "]    - Padding character
 * @return {string} Padded string
 */
export function padLeft(str: string, len: number, padChar: string = " "): string {
    if (str.length >= len) {
        return str;
    }

    // Only allow single character padding
    if (padChar.length > 1) {
        throw new Error("Specify only one character");
    }

    // Should be at least one character
    if (padChar.length === 0) {
        throw new Error("Specify at least one character");
    }

    return padChar.repeat(len - str.length) + str;
}

/**
 * Trims a given line based on the terminal's width. It also replaces tabs with spaces.
 *
 * @param {string} str - string to be trimmed
 * @return {string} Trimmed string
 */
export function trimLineToTerminalWidth(str: string): string {
    const terminalWidth = process.stdout.columns - 1;
    const ellipsisLength = 3;
    str = str.replace(/\t/g, " ");

    // No need to trim this string
    // The second part of this if statement will only happen IFF the terminal width is also less than (or equal to) the ellipsisLength (e.g. 3)
    if (stringWidth(str) < terminalWidth || stringWidth(str) <= ellipsisLength) {
        return str;
    }

    return `${str.slice(0, -1 * (stringWidth(str) - terminalWidth + ellipsisLength))}...`;
}

/**
 * Produces formatted context of a given error in the specified location of the contents
 *
 * NOTE:
 *      The lineIndex is zero-based indexed
 *
 * @param {string | string[]} content   - Indicates the contents or the file path to the file containing the contents to get the error from
 * @param {number} lineIndex            - Zero-basd indexed location of the error to be printed
 * @return {string} Few lines (separated by \n) containing error context
 */
export function getErrorContext(content: string | string[], lineIndex: number): string {
    let retValue = "";

    if (lineIndex < 0) {
        // No error location provided
        return retValue;
    }

    let lines = content;
    if (typeof content === "string") {
        // Try reading from a file
        if (fs.existsSync(content)) {
            lines = fs.readFileSync(content).toString().split(/\r?\n/);
        } else {
            throw new Error(`File ${content} not found!`);
        }
    }

    // Get the possible padding length for the line numbers
    const maxLineIndexLength = Math.max((lineIndex + 1).toString().length, (lineIndex + 2).toString().length);

    // Gather the line before in case we are able to get it
    if (lineIndex > 0) {
        retValue += trimLineToTerminalWidth(`\n   ${padLeft(lineIndex.toString(), maxLineIndexLength)} |  ${lines[lineIndex - 1]}`);
    }

    // Format the current line of the
    retValue +=
        trimLineToTerminalWidth(`\n > ${TextUtils.chalk.red(padLeft((lineIndex + 1).toString(), maxLineIndexLength))} |  ${lines[lineIndex]}`);

    // Gather the next line in case we are able to get it
    if (lineIndex + 1 < lines.length) {
        retValue += trimLineToTerminalWidth(`\n   ${padLeft((lineIndex + 2).toString(), maxLineIndexLength)} |  ${lines[lineIndex + 1]}`);
    }

    return retValue;
}

/**
 * This is a simple sleep function to help when waiting for a certain period of
 * time before continuing.
 *
 * @param {number} [ms=1000] The number of milliseconds to sleep.
 *
 * @returns {Promise<void>} Resolves after the specified time is up.
 *
 * @example <caption>Sleeping within code</caption>
 *
 * async function doStuff() {
 *   await CoreUtils.sleep(10000);
 *   console.log("You will see this message 10 seconds after function was entered");
 * }
 */
export function sleep(ms = 1000): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Utility function to read complete standard in contents from a handler
 * Can be awaited from async methods
 */
export function readStdin(): Promise<Buffer> {
    return new Promise<Buffer>((resolve, reject) => {
        let stdinContent: Buffer = Buffer.from([]);
        const stream = process.stdin;
        const log = Logger.getAppLogger();
        stream.resume();

        stream.on("data", (chunk: Buffer) => {
            log.trace(`Read data from stdin: ${chunk.toString()}`);
            log.debug(`Read ${chunk.length} bytes of data from stdin`);
            stdinContent = Buffer.concat([stdinContent, chunk]);
        });

        stream.once("end", () => {
            log.info("Finished reading stdin");
            resolve(stdinContent);
        });

        stream.on("error", (error: Error) => {
            const stdinReadError: IImperativeError = {
                msg: "Error encountered while reading from stdin",
                causeErrors: error,
                additionalDetails: (error == null) ? undefined : error.message
            };
            reject(stdinReadError);
        });
    });
}

/**
 * Execute multiple promises in a pool with a maximum number of promises
 * executing at once
 * @param poolLimit - limit of how many promises should execute at once
 * @param array - array of objects to convert to promises with iteratorFn
 * @param iteratorFn - the function that turns an entry in the array into a promise
 */
export function asyncPool(poolLimit: number, array: any[],
    iteratorFn: (item: any, array: any[]) => Promise<any>): Promise<any> {
    let i = 0;
    const ret: any[] = [];
    const executing: any[] = [];
    const enqueue: any = () => {
        if (i === array.length) {
            return Promise.resolve();
        }
        const item = array[i++];
        const p = Promise.resolve().then(() => iteratorFn(item, array));
        ret.push(p);
        const e: any = p.then(() => executing.splice(executing.indexOf(e), 1));
        executing.push(e);
        let r = Promise.resolve();
        if (executing.length >= poolLimit) {
            r = Promise.race(executing);
        }
        return r.then(() => enqueue());
    };
    return enqueue().then(() => Promise.all(ret));
}
