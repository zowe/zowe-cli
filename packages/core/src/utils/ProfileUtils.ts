/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeConfig, IProfileLoaded, CliProfileManager, IProfile, Logger, ImperativeError } from "@zowe/imperative";
import * as path from "path";

/**
 * Retrieves the Zowe CLI home directory. In the situation Imperative has
 * not initialized it we mock a default value.
 * @returns {string} - Returns the Zowe home directory
 */
export function getZoweDir(): string {
    if (ImperativeConfig.instance.loadedConfig == null) {
        ImperativeConfig.instance.loadedConfig = {
            defaultHome: path.join(require("os").homedir(), ".zowe"),
            envVariablePrefix: "ZOWE"
        };
    }
    return ImperativeConfig.instance.cliHome;
}

/**
 * Loads default z/OSMF profile. The profile must have already been
 * created using Zowe CLI, and not rely on base profiles
 *
 * @param {string} profileType - The name of the profile type
 * @param {boolean} mergeWithBase - Whether or not to merge with the base profile
 * @returns {IProfile} - The default profile (or merged profile, if requested)
 */
export async function getDefaultProfile(profileType: string, mergeWithBase?: boolean) {

    const profileRootDir: string = path.join(getZoweDir(), "profiles");
    const logger: Logger = Logger.getImperativeLogger();
    let profileManager: CliProfileManager;
    let profileLoaded: IProfileLoaded;

    try {
        // Create the profile manager
        profileManager = new CliProfileManager({
            profileRootDirectory: profileRootDir,
            type: profileType
        });

        // Load the profile using the profile manager
        profileLoaded = await profileManager.load({
            loadDefault: true
        });
    } catch (err) {
        logger.warn(err.message);
    }

    // Throw an error if there was no default profile found and no base profile requested
    if ((!profileLoaded || !profileLoaded.profile) && !mergeWithBase) {
        throw new ImperativeError({msg: `Failed to load default profile of type "${profileType}"`});
    }

    // Give the profile back as-is if the profile is not to be merged with the base profile
    if (mergeWithBase === false) {
        return profileLoaded.profile;
    }

    let baseProfileManager: CliProfileManager;
    let baseProfileLoaded: IProfileLoaded;

    try {
        baseProfileManager = new CliProfileManager({
            profileRootDirectory: profileRootDir,
            type: "base"
        });

        baseProfileLoaded = await baseProfileManager.load({
            loadDefault: true
        });
    } catch (err) {
        logger.warn(err.message);
    }

    // Return service profile if there was no base profile found and service profile existed.
    // Return base profile if there was no service profile found and base profile existed.
    // If both exist, combine. Otherwise, error
    if ((!baseProfileLoaded || !baseProfileLoaded.profile) && (profileLoaded && profileLoaded.profile)) {
        return profileLoaded.profile;
    } else if (baseProfileLoaded && baseProfileLoaded.profile && (!profileLoaded || !profileLoaded.profile)) {
        return baseProfileLoaded.profile;
    } else if ((!baseProfileLoaded || !baseProfileLoaded.profile) && (!profileLoaded || !profileLoaded.profile)) {
        throw new ImperativeError({msg: `Failed to load default profiles of types "${profileType}" and "base"`});
    }

    const combinedProfile: IProfile = JSON.parse(JSON.stringify(baseProfileLoaded.profile));

    for (const propertyName in profileLoaded.profile) {
        if (profileLoaded.profile[propertyName] != null) {
            combinedProfile[propertyName] = profileLoaded.profile[propertyName];
        }
    }

    return combinedProfile;
}