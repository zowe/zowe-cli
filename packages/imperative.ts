/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

// WIP Imperative version of Brightside
import { IImperativeConfig } from "@brightside/imperative";
import { Constants } from "./Constants";
import { ZosmfSession } from "./zosmf";
import {
    TSO_OPTION_ACCOUNT,
    TSO_OPTION_CHAR_SET,
    TSO_OPTION_CODE_PAGE,
    TSO_OPTION_COLUMNS,
    TSO_OPTION_LOGON_PROCEDURE,
    TSO_OPTION_REGION_SIZE,
    TSO_OPTION_ROWS
} from "./zostso/src/cli/constants/ZosTso.constants";

const config: IImperativeConfig = {
    productDisplayName: Constants.DISPLAY_NAME,
    commandModuleGlobs: ["**/cli/*.definition!(.d).*s"],
    rootCommandDescription: Constants.DESCRIPTION,
    defaultHome: Constants.HOME_DIR,
    envVariablePrefix: Constants.ENV_PREFIX,
    logging: {
        appLogging: {
            logFile: Constants.LOG_LOCATION
        },
    },
    profiles: [
        {
            type: "zosmf",
            schema: {
                type: "object",
                title: "z/OSMF Profile",
                description: "z/OSMF Profile",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_HOST,
                    },
                    port: {
                        type: "number",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PORT,
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_USER,
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_PASSWORD,
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_REJECT_UNAUTHORIZED,
                    },
                    basePath: {
                        type: "string",
                        optionDefinition: ZosmfSession.ZOSMF_OPTION_BASE_PATH,
                    },
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "zos123 --host zos123 --port 1443 --user ibmuser --password myp4ss",
                    description: "Create a zosmf profile called 'zos123' to connect to z/OSMF at host zos123 and port 1443"
                },
                {
                    options: "zos124 --host zos124 --user ibmuser --password myp4ss --reject-unauthorized false",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                },
                {
                    options: "zosAPIML --host zosAPIML --port 2020 --user ibmuser --password myp4ss --reject-unauthorized false --base-path basePath",
                    description: "Create a zosmf profile called 'zos124' to connect to z/OSMF at the host zos124 (default port - 443) " +
                        "and allow self-signed certificates"
                }
            ],
            updateProfileExamples: [
                {
                    options: "zos123 --user newuser --password newp4ss",
                    description: "Update a zosmf profile named 'zos123' with a new username and password"
                }
            ]
        },
        {
            type: "tso",
            schema: {
                type: "object",
                title: "TSO Profile",
                description: "z/OS TSO/E User Profile",
                properties: {
                    account: {
                        type: "string",
                        optionDefinition: TSO_OPTION_ACCOUNT,
                    },
                    characterSet: {
                        type: "string",
                        optionDefinition: TSO_OPTION_CHAR_SET,
                    },
                    codePage: {
                        type: "string",
                        optionDefinition: TSO_OPTION_CODE_PAGE,
                    },
                    columns: {
                        type: "number",
                        optionDefinition: TSO_OPTION_COLUMNS,
                    },
                    logonProcedure: {
                        type: "string",
                        optionDefinition: TSO_OPTION_LOGON_PROCEDURE,
                    },
                    regionSize: {
                        type: "number",
                        optionDefinition: TSO_OPTION_REGION_SIZE,
                    },
                    rows: {
                        type: "number",
                        optionDefinition: TSO_OPTION_ROWS,
                    },
                },
                required: ["account"],
            },
            createProfileExamples: [
                {
                    description: "Create a tso profile called 'myprof' with default settings and JES accounting information of 'IZUACCT'",
                    options: "myprof -a IZUACCT"
                },
                {
                    description: "Create a tso profile called 'largeregion' with a region size of 8192, a logon procedure of MYPROC, and " +
                        "JES accounting information of '1234'",
                    options: "largeregion -a 1234 --rs 8192"
                }
            ],
            updateProfileExamples: [
                {
                    description: "Update a tso profile called myprof with new JES accounting information",
                    options: "myprof -a NEWACCT"
                }
            ]
        },
    ]
};
module.exports = config;
