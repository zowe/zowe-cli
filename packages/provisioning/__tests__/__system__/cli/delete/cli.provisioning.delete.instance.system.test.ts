/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Imperative, Session } from "@zowe/imperative";
import { runCliScript } from "../../../../../../__tests__/__src__/TestUtils";
import { ITestEnvironment } from "../../../../../../__tests__/__src__/environment/doc/response/ITestEnvironment";
import { TestEnvironment } from "../../../../../../__tests__/__src__/environment/TestEnvironment";
import { ITestSystemSchema } from "../../../../../../__tests__/__src__/properties/ITestSystemSchema";
import { TestProperties } from "../../../../../../__tests__/__src__/properties/TestProperties";
import { PerformAction, ProvisioningConstants } from "../../../../";
import { ProvisioningTestUtils } from "../../../__resources__/utils/ProvisioningTestUtils";
import { ITestZosmfSchema } from "../../../../../../__tests__/__src__/properties/ITestZosmfSchema";

// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;
let TEST_ENVIRONMENT_NO_PROF: ITestEnvironment;
let REAL_SESSION: Session;
let defaultSystem: ITestPropertiesSchema;
let templateName: string;
let instance: any;
let instanceID: string;
let instanceName: string;

describe("provisioning delete instance", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "provisioning_delete_instance",
            tempProfileTypes: ["zosmf", "tso"]
        });

        defaultSystem = TEST_ENVIRONMENT.systemTestProperties;
        templateName = templateName = TEST_ENVIRONMENT.systemTestProperties.provisioning.templateName;
        REAL_SESSION = TestEnvironment.createZosmfSession(TEST_ENVIRONMENT);

        // Provision the template to have an instance to delete
        instance = await ProvisioningTestUtils.getProvisionedInstance(REAL_SESSION, ProvisioningConstants.ZOSMF_VERSION, templateName);
        Imperative.console.info(`Provisioned instance: ${instance["external-name"]}`);
        instanceID = instance["object-id"];

        // Deprovision the instance
        instance = await PerformAction.doProvisioningActionCommon(REAL_SESSION, ProvisioningConstants.ZOSMF_VERSION,
            instanceID, ProvisioningTestUtils.ACTION_DEPROV);
        Imperative.console.info(`Deprovision of the instance started, action-id: ${instance["action-id"]}`);
        // Wait until instance state is 'deprovisioned'
        instance = await ProvisioningTestUtils.waitInstanceState(REAL_SESSION, ProvisioningConstants.ZOSMF_VERSION,
            instanceID, ProvisioningTestUtils.STATE_DEPROV);
        instanceName = instance["external-name"];
        Imperative.console.info(`Deprovisioned instance: ${instanceName}`);
    }, ProvisioningTestUtils.MAX_TIMEOUT_TIME);

    it("should successfully issue the command", async () => {
        const response = runCliScript(__dirname + "/__scripts__/instance/delete_instance_success.sh", TEST_ENVIRONMENT,
            [instanceName]);
        expect(response.stderr.toString()).toBe("");
        expect(response.status).toBe(0);
        expect(response.stdout).toMatchSnapshot();
    }, ProvisioningTestUtils.MAX_CLI_TIMEOUT);


    describe("without profiles", () => {
        let zOSMF: ITestZosmfSchema;

        // Create a separate test environment for no profiles
        beforeAll(async () => {
            TEST_ENVIRONMENT_NO_PROF = await TestEnvironment.setUp({
                testName: "provisioning_delete_instance_no_profile",
            });
            zOSMF = TEST_ENVIRONMENT_NO_PROF.systemTestProperties.zosmf;

            // Provision the template to have an instance to delete
            instance = await ProvisioningTestUtils.getProvisionedInstance(REAL_SESSION, ProvisioningConstants.ZOSMF_VERSION, templateName);
            Imperative.console.info(`Provisioned instance: ${instance["external-name"]}`);
            instanceID = instance["object-id"];

            // Deprovision the instance
            instance = await PerformAction.doProvisioningActionCommon(REAL_SESSION, ProvisioningConstants.ZOSMF_VERSION,
                instanceID, ProvisioningTestUtils.ACTION_DEPROV);
            Imperative.console.info(`Deprovision of the instance started, action-id: ${instance["action-id"]}`);
            // Wait until instance state is 'deprovisioned'
            instance = await ProvisioningTestUtils.waitInstanceState(REAL_SESSION, ProvisioningConstants.ZOSMF_VERSION,
                instanceID, ProvisioningTestUtils.STATE_DEPROV);
            instanceName = instance["external-name"];
            Imperative.console.info(`Deprovisioned instance: ${instanceName}`);
        }, ProvisioningTestUtils.MAX_TIMEOUT_TIME);

        it("should successfully delete an instance without profile", async () => {
            const response = runCliScript(__dirname + "/__scripts__/instance/delete_instance_fully_qualified.sh",
                TEST_ENVIRONMENT_NO_PROF,
                [
                    instanceName,
                    zOSMF.host,
                    zOSMF.port,
                    zOSMF.user,
                    zOSMF.pass
                ]);
            expect(response.stderr.toString()).toBe("");
            expect(response.status).toBe(0);
            expect(response.stdout).toMatchSnapshot();
        }, ProvisioningTestUtils.MAX_CLI_TIMEOUT);
    });
    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT_NO_PROF);
    });
});
