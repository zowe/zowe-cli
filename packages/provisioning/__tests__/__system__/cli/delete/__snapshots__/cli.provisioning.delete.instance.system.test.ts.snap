// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provisioning delete instance should successfully issue the command 1`] = `
Object {
  "data": Array [
    84,
    104,
    101,
    32,
    105,
    110,
    115,
    116,
    97,
    110,
    99,
    101,
    32,
    119,
    97,
    115,
    32,
    115,
    117,
    99,
    99,
    101,
    115,
    115,
    102,
    117,
    108,
    108,
    121,
    32,
    100,
    101,
    108,
    101,
    116,
    101,
    100,
    10,
  ],
  "type": "Buffer",
}
`;

exports[`provisioning delete instance without profiles should successfully delete an instance without profile 1`] = `
Object {
  "data": Array [
    84,
    104,
    101,
    32,
    105,
    110,
    115,
    116,
    97,
    110,
    99,
    101,
    32,
    119,
    97,
    115,
    32,
    115,
    117,
    99,
    99,
    101,
    115,
    115,
    102,
    117,
    108,
    108,
    121,
    32,
    100,
    101,
    108,
    101,
    116,
    101,
    100,
    10,
  ],
  "type": "Buffer",
}
`;
